mopidy_cd/backend.py

```diff
@@ -1,65 +1,81 @@
+from __future__ import unicode_literals
+
+import logging
+
 import pykka
 
 from mopidy import backend
-from mopidy.models import Ref, Track, Album
-from . import cdrom
-import logging
+from mopidy.models import Album, Artist, Image, Ref, Track
+
+from cdrom import CdRom
+
 
 logger = logging.getLogger(__name__)
 
+URI_PREFIX = 'cd:/'
+
 
 class CdBackend(pykka.ThreadingActor, backend.Backend):
+
     uri_schemes = ['cd']
 
     def __init__(self, config, audio):
         super(CdBackend, self).__init__()
-        self.audio = audio
-        self.cdrom = cdrom.Cdrom()
-
+        self.cdrom = CdRom()
         self.library = CdLibrary(backend=self)
-        self.playback = CdPlaybackProvider(audio=audio, backend=self)
+        self.playback = CdPlayback(audio=audio, backend=self)
 
 
 class CdLibrary(backend.LibraryProvider):
-    root_directory = Ref.directory(uri='cd:root', name='Cd')
+
+    root_directory = Ref.directory(uri=URI_PREFIX + 'root', name='CD')
 
     def browse(self, uri):
+        def make_track_ref(track):
+            return Ref.track(
+                uri=URI_PREFIX + str(track.number), name=track.title
+            )
+
         self.refresh()
-        results = []
-        if not uri == 'cd:root':
-            return results
-        logger.debug('Cdrom backend %s', self.backend)
-        tracks = self.backend.cdrom.tracks
-        logger.debug('Cdrom: in browse found %d tracks', len(tracks))
-        for (seq, (number, name, duration, albumtitle, genre, year)) in enumerate(tracks, 1):
-            results.append(Ref.track(uri='cd:/' + str(seq), name=name))
-        return results
 
-    def refresh(self, uri=None):
-        self.backend.cdrom.refresh()
+        return map(make_track_ref, self.backend.cdrom.disc.tracks)
 
     def lookup(self, uri):
-        logger.debug('Cdrom: track selected')
-        i = int(uri.lstrip("cd:/")) - 1
-        logger.debug('Cdrom: track %s selected', i)
-        (number, name, duration, albumtitle,
-         genre, year) = self.backend.cdrom.tracks[i]
-        album = Album(name=albumtitle)
-        return [Track(uri=uri,
-                      name=name,
-                      length=int(duration) * 1000,
-                      genre=genre,
-                      date=year,
-                      album=album)]
-
-    def search(self, query=None, uris=None, exact=False):
-        return
-
-class CdPlaybackProvider(backend.PlaybackProvider):
-
-    def change_track(self, track):
-        logger.debug('Cdrom: playing track %s', track)
-        return super(CdPlaybackProvider, self).change_track(track)
+        def make_artist(artist_tuple):
```

Comment:
Wouldn't it make more sense for `make_artist` to be a global function, and not nested in here? This is not a closure using some variable of the enclosed scope.

Reply:
It was basically done to narrow the visibility scope of `make_artist` fn as it's not usable anywhere else outside `lookup` fn, but I don't have a strong preference here and happy to change 😄

Reply:
I guess nested functions are a bit slower, and if there is no need for that I think it is better to have global ones.

Reply:
Moved to a static method